File: patterns/structural_flyweight/structural_flyweight_webref.js
class Auto {
	constructor(model) {
		this.model = model;
	}
}

class AutoFactory {
	constructor(name) {
		this.models = {};
	}

	create(name) {
		let model = this.models[name];
        if (model) return model;
        console.count('model');
		this.models[name] = new Auto(name);
		return this.models[name];
    }
    
    getModels() {
        console.table(this.models);
    }
};

// --
const factory = new AutoFactory();

const bmw = factory.create('bmw');
const audi = factory.create('audi');
const tesla = factory.create('tesla');
const teslaBlack = factory.create('tesla');

console.log(factory.getModels());

File: patterns/structural_flyweight/structural_flyweight_webref.py
class Auto:
  def __init__(self, model):
    self.model = model


class AutoFactory:
  def __init__(self):
    self.models = {}
    self.count = 0

  def create(self, name):
    model = self.models.get(name)

    if model:
      return model

    self.count = self.count + 1

    self.models[name] = Auto(name)
    return self.models[name]

  def get_models(self):
    print("Count autos: %s" % self.count)
    for model in self.models:
      print(model)

if __name__ == "__main__":
  factory = AutoFactory()

  bmw = factory.create('bmw')
  audi = factory.create('audi')
  tesla = factory.create('tesla')
  teslaBlack = factory.create('tesla')

  factory.get_models()
    